DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h

```diff
@@ -257,16 +263,50 @@ class TTTrack_TrackWord {
                     ap_uint<TrackBitWidths::kMVAQualitySize> mvaQuality,
                     ap_uint<TrackBitWidths::kMVAOtherSize> mvaOther);
 
+  void testDigitizationScheme();
+
 private:
   // ----------private member functions --------------
+  // Return the number of set bits in the binary representation of 'value'
+  unsigned int countSetBits(unsigned int value) const {
+    unsigned int count = 0;
+    while (value) {
+      value &= (value - 1);
+      count++;
+    }
+    return count;
+  }
+
+  // Return the number of set bits in the given range
+  unsigned int countSetBitsInGivenRange(unsigned int value, unsigned int lsb, unsigned int msb) const {
+    // Calculate a number 'num' having 'msb' number of bits and bits in the range 'lsb' to 'msb' are the only set bits
+    // Note: 'lsb' and 'msb' are ordered as follows 1 <= lsb <= msb <= <number of bits in value>
+    int num = ((1 << msb) - 1) ^ ((1 << (lsb - 1)) - 1);
+
+    // returns number of set bits in the range 'lsb' to 'msb' in 'value'
+    return countSetBits(value & num);
+  }
+
   unsigned int digitizeSignedValue(double value, unsigned int nBits, double lsb) const {
-    unsigned int digitized_value = std::abs(std::floor(value / lsb));
-    unsigned int digitized_maximum = (1 << (nBits - 1)) - 1;  // The remove 1 bit from nBits to account for the sign
-    if (digitized_value > digitized_maximum)
-      digitized_value = digitized_maximum;
-    if (value < 0)
-      digitized_value = (1 << nBits) - digitized_value;  // two's complement encoding
-    return digitized_value;
+    // Digitize the incoming value
+    unsigned int digitizedValue = std::abs(std::floor(value / lsb));
+
+    // Calculate the maxmum possible positive value given an output of nBits in size
+    unsigned int digitizedMaximum = (1 << (nBits - 1)) - 1;  // The remove 1 bit from nBits to account for the sign
+    unsigned int digitizedMinimum = (1 << (nBits - 1));
+
+    // Saturate the digitized value
```

Comment:
If the goal of this is to rescue numbers that are slightly outside the digi range, I don't think it works. If digi range is from -4 to +3 (lsb=1,nbits=3), then it digitizedMaximum=4, so this code will only change the value of twosValue of digitizedValue (which is abs of digi value) is 5 or more. So if numerical errors had caused the input digi to be +4, it would still lie outside the allowed range after this code.

Reply:
Hi @tomalin, I see what you mean. What do you think about changing the conditional to:
```
    if ( (value >= 0) && (digitizedValue > digitizedMaximum) ) {
      twosValue = digitizedMaximum;
    }
    else if( (value < 0) && (digitizedValue > digitizedMinimum) ) {
      twosValue = digitizedMinimum;
    }
```
Using the input values of `[3,4,5,-4,-5]` with `lsb=1` and `nBits=3` I find the correct behavior:
```
value ==> twosValue
3 ==> 3 // neither conditional is true
4 ==> 3 // 1st conditional evaluates to true
5 ==> 3 // 1st conditional evaluates to true
-4 ==> 4 // neither conditional is true
-5 ==> 4 // 2nd conditional evaluates to true
```
Do you agree that this would satisfy the issue?

Reply:
I agree that would be correct. Though if the range checking is now asymmetric, I'd suggest the code would be clearer by removing the abs() from https://github.com/cms-l1t-offline/cmssw/blob/e4bbee047752a8b5f9ad1bdbc6d7c78950b6cb68/DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h#L300 , so you work with conventional digi values . i.e.

int digitizedValue = std::floor(value / lsb);
int digitizedMaximum = (1 << (nBits - 1)) - 1;  
int digitizedMinimum = -(digitizedMaximum + 1);
if (digitizedValue > digitizedMaximum) digitizedValue = digitizedMaximum;
if (digitizedValue < digitizedMinimum) digitizedValue = digitizedMinimum;
int twosValue;
if (digitizedValue > 0) {
   twosValue = digitizedValue;
} else {
  twosValue = digitizedValue + (1 << nBits);
}

Reply:
I second @tomalin's suggestion.

Depending on taste, you could replace the two if's with std::min and std::max, or better if c++17 is allowed in CMSSW  then [std::clamp](https://en.cppreference.com/w/cpp/algorithm/clamp), i.e. `digitizedValue = std::clamp(digitizedValue, digitizedMinimum, digitizedMaximum)`

Reply:
I have integrated both of your suggestions.